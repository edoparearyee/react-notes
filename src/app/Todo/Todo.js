import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import './Todo.scss';

class Todo extends Component {
  render() {
    const todoClass = classNames({
      Todo: true,
      'Todo--Complete': this.props.complete,
    });
    return (
      <div className={todoClass}>
        <input
          className="Todo__Checkbox"
          type="checkbox"
          checked={this.props.complete}
          onChange={e => this.handleCompleteChange(e)}
        />
        <p
          className="Todo__Title"
          ref={c => (this._titleEl = c)}
          contentEditable={!this.props.complete}
          onInput={e => this.handleTitleChange(e)}
        >
          {this.props.title}
        </p>
      </div>
    );
  }

  handleCompleteChange(event) {
    this.props.onChange({
      id: this.props.id,
      complete: event.target.checked,
      title: this.props.title,
    });
  }

  handleTitleChange(event) {
    this.props.onChange({
      id: this.props.id,
      complete: this.props.complete,
      title: event.target.innerText,
    });
  }
}

Todo.propTypes = {
  id: PropTypes.number.isRequired,
  complete: PropTypes.bool,
  title: PropTypes.string.isRequired,
  onChange: PropTypes.func,
};

export default Todo;
