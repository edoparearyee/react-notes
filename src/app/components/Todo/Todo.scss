@import "../../../scss/variables";

.Todo {
  align-items: center;
  display: flex;
  font-family: $font-family-sans-serif;
}

.Todo__Checkbox {
  margin-right: 0.5rem;
}

.Todo__Title {
  outline: none;

  .Todo--Complete & {
    text-decoration: line-through;
  }
}
