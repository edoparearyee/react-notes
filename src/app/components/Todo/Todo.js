import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import './Todo.scss';

class Todo extends Component {
  render() {
    const todoClass = classNames({
      Todo: true,
      'Todo--Complete': this.props.complete,
    });
    return (
      <li className={todoClass}>
        <input
          className="Todo__Checkbox"
          ref={c => (this._checkboxEl = c)}
          type="checkbox"
          value={this.props.complete}
          onInput={e => this.handleToggle(e)}
        />
        <p
          className="Todo__Title"
          ref={c => (this._titleEl = c)}
          contentEditable={!this.props.complete}
          onInput={e => this.handleInput(e)}
        >
          {this.props.title}
        </p>
      </li>
    );
  }

  shouldComponentUpdate(nextProps) {
    return (
      nextProps.title !== this._titleEl.innerText ||
      nextProps.complete !== this.props.complete
    );
  }

  handleToggle() {
    this.props.onToggle({
      id: this.props.id,
    });
  }

  handleInput(event) {
    this.props.onChange({
      id: this.props.id,
      title: event.target.innerText,
    });
  }
}

Todo.propTypes = {
  id: PropTypes.string.isRequired,
  complete: PropTypes.bool,
  title: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  onToggle: PropTypes.func,
};

export default Todo;
