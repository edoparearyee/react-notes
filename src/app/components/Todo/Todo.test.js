import React from 'react';
import ReactDOM from 'react-dom';
import Todo from './Todo';

const todo = {
  id: 0,
  title: 'Culpa excepteur nisi fugiat',
  complete: false,
};

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Todo title={todo.title} complete={todo.todo} />, div);
  ReactDOM.unmountComponentAtNode(div);
});
