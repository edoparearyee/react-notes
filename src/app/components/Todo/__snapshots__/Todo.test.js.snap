// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo renders component 1`] = `
<li
  className="Todo"
>
  <input
    checked={false}
    className="Todo__Checkbox"
    onChange={[Function]}
    onInput={[Function]}
    type="checkbox"
  />
  <p
    className="Todo__Title"
    contentEditable={true}
    onInput={[Function]}
    suppressContentEditableWarning={true}
  >
    Culpa excepteur nisi fugiat
  </p>
  <button
    className="Todo__Delete"
    onClick={[Function]}
    type="button"
  >
    Delete
  </button>
</li>
`;
